pints/_mcmc/_slice_stepout.py

```diff
@@ -0,0 +1,430 @@
+    6. ``K = (m - 1) - J``
+    6. while ``J > 0`` and ``y < f(L)``, ``L = L - w`` and ``J = J - 1``
+    7. while ``K > 0`` and ``y < f(R)``, ``R = R + w`` and ``K = K - 1``
+
+    Intuitively, the interval ``I`` is estimated by expanding the initial
+    interval by a width ``w`` in each direction until both edges fall outside
+    the slice, or until a pre-determined limit is reached. The parameters
+    ``m`` (an integer, which determines the limit of slice size) and
+    ``w`` (the estimate of typical slice width) are hyperparameters.
+
+    To sample from the interval ``I = (L, R)``, such that the sample
+    ``x`` satisfies ``y < f(x)``, we use the ``Shrinkage`` procedure, which
+    reduces the size of the interval after rejecting a trial point,
+    as defined in [1] Fig. 5. pp.716. This algorithm consists of the
+    following steps:
+
+    1. ``\bar{L} = L`` and ``\bar{R} = R``
+    2. Repeat:
+        a. ``U \sim uniform(0, 1)``
+        b. ``x_1 = \bar{L} + U (\bar{R} - \bar{L})``
+        c. if ``y < f(x_1)`` accept ``x_1`` and exit loop,
+           else
+            if ``x_1 < x_0``, ``\bar{L} = x_1``
+            else ``\bar{R} = x_1``
+
+    Intuitively, we uniformly sample a trial point from the interval ``I``,
+    and subsequently shrink the interval each time a trial point is rejected.
+
+    To avoid floating-point underflow, we implement the suggestion advanced
+    in [1] pp.712. We use the log pdf of the un-normalised posterior
+    (``g(x) = log(f(x))``) instead of ``f(x)``. In doing so, we use an
+    auxiliary variable ``z = log(y) = g(x0) − \epsilon``, where
+    ``\epsilon \sim \text{exp}(1)`` and define the slice as
+    S = {x : z < g(x)}.
+
+    [1] Neal, R.M., 2003. Slice sampling. The annals of statistics, 31(3),
+    pp.705-767.
+    """
+
+    def __init__(self, x0, sigma0=None):
+        super(SliceStepoutMCMC, self).__init__(x0, sigma0)
+
+        # Set initial state
+        self._x0 = np.asarray(x0, dtype=float)
+        self._running = False
+        self._ready_for_tell = False
+
+        # Iterations monitoring
+        self._mcmc_iteration = 0
+
+        # Current sample, log_pdf of the current sample
+        self._current = None
+        self._current_log_pdf = None
+
+        # Current log_y used to define the slice
+        self._current_log_y = None
+
+        # Current proposed sample
+        self._proposed = None
+
+        # Default initial interval width w used in the Stepout procedure
+        # to expand the interval
+        self._w = np.ones(len(self._x0))
+
+        # Default integer limiting the size of the interval to ``m*w```
+        self._m = 50
+
+        # Flag to initialise the expansion of the interval ``I=(L,R)``
+        self._first_expansion = False
+
+        # Flag indicating whether the interval expansion is concluded
+        self._interval_found = False
+
+        # Number of steps used for expanding the interval ``I=(L,R)``
+        self._j = None
```

Comment:
I think these could be named using easier to interpret variables. For example self._l_count and self._r_count or something?

Reply:
@ben18785 I am happy to change the names of such variables if you prefer - I used "j" and "k" since these are the names used in Neal's paper's pseudocode. I just thought that keeping the same parameters' names would allow the user to understand better the code by referring to the paper's pseudocode. Would you prefer to change the names of these variables anyway?